File: 41.缺失的第一个正数.cpp
/*
 * @lc app=leetcode.cn id=41 lang=cpp
 *
 * [41] 缺失的第一个正数
 */
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        int i;
        int n = nums.size();
        int * B = (int *) malloc(sizeof(int) * n);
        memset(B, 0, sizeof(int) * n);
        for (i = 0; i < n; ++i)
        {
            if (nums[i] > 0 && nums[i] <= n)
                B[nums[i] - 1] = 1;
        }
        for (i = 0; i < n; i++)
        {
            if (B[i] == 0)
                break;
        }
        return i + 1;
    }
};

File: 41.缺失的第一个正数.java
/*
 * @lc app=leetcode.cn id=41 lang=java
 *
 * [41] 缺失的第一个正数
 */

// @lc code=start
class Solution {
    // 置换法：把数组还原成按顺序排列的数组，
    public int firstMissingPositive(int[] nums) {
        int n = nums.length;
        for (int i = 0; i < n; i++) {
            // 用 while 直到数字换到了正确的位置。
            // 也就是 nums[i] = i + 1
            while (nums[i] > 0 && nums[i] <= n && nums[nums[i] - 1] != nums[i]) {
                int temp = nums[nums[i] - 1];
                nums[nums[i] - 1] = nums[i];
                nums[i] = temp;
            }
        }
        for (int i = 0; i < n; i++) {
            if (nums[i] != i + 1) {
                return i + 1;
            }
        }
        return n + 1;
    }
}
// @lc code=end

